dask_cuda/device_host_file.py

```diff
@@ -1,26 +1,71 @@
 from zict import Buffer, File, Func
 from zict.common import ZictBase
-from distributed.protocol import deserialize_bytes, serialize_bytes
+from distributed.protocol import (
+    deserialize,
+    deserialize_bytes,
+    serialize,
+    serialize_bytelist,
+    Serialized,
+)
+from dask.sizeof import sizeof
+from distributed.utils import nbytes
 from distributed.worker import weight
 
 from functools import partial
+from numba import cuda
 import os
 
 from .is_device_object import is_device_object
 
 
+# Register sizeof for Numba DeviceNDArray while Dask doesn't add it
+if not hasattr(sizeof, "register_numba"):
+
+    @sizeof.register_lazy("numba")
+    def register_numba():
+        import numba
+
+        @sizeof.register(numba.cuda.cudadrv.devicearray.DeviceNDArray)
+        def sizeof_numba_devicearray(x):
+            return int(x.nbytes)
+
+
+# Register sizeof for Serialized while Dask doesn't add it
+if not hasattr(sizeof, "register_serialized"):
+
+    @sizeof.register(Serialized)
+    def register_serialized(obj):
+        return sum(map(nbytes, obj.frames))
+
+
+def device_to_host(obj: object) -> Serialized:
+    header, frames = serialize(obj, serializers=["cuda"])
+    is_cuda = [hasattr(f, "__cuda_array_interface__") for f in frames]
+    frames = [
+        cuda.as_cuda_array(f).copy_to_host() if ic else f
+        for ic, f in zip(is_cuda, frames)
+    ]
+    header = {"sub-header": header, "is-cuda": is_cuda}
+    return Serialized(header, frames)
+
+
+def host_to_device(s: Serialized) -> object:
+    is_cuda = s.header["is-cuda"]
+    header = s.header["sub-header"]
+    frames = [cuda.to_device(f) if ic else f for ic, f in zip(is_cuda, s.frames)]
+    return deserialize(header, frames, serializers=["cuda"])
+
+
 def _serialize_if_device(obj):
-    """ Serialize an object if it's a device object """
     if is_device_object(obj):
-        return serialize_bytes(obj, on_error="raise")
+        return device_to_host(obj)
     else:
         return obj
 
 
 def _deserialize_if_device(obj):
-    """ Deserialize an object if it's an instance of bytes """
-    if isinstance(obj, bytes):
-        return deserialize_bytes(obj)
+    if isinstance(obj, Serialized):
+        return host_to_device(obj)
```

Comment:
Will the else condition of this check ever occur?  Are there cases when place device objects into host without serializing them?  If this shouldn't occur then perhaps we could turn this check into an assertion?

Reply:
> Will the else condition of this check ever occur?

I expect so.

> Are there cases when place device objects into host without serializing them?

This case shouldn't exist.

> If this shouldn't occur then perhaps we could turn this check into an assertion?

 It should always be the case that a device object will be an instance of `Serialized` and we would want to move it back to device. However, as I was trying to confirm this, none of the tests seem to call `_deserialize_if_device` at any point, which is weird, I'm now investigating it.